### ISSUE TITLE
Symbolic functions in the ntheory module

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#16244]
Many functions in the ntheory module only accept integer arguments. But they could be useful accepting symbolic arguments as well. For instance, many infinite products cannot currently be represented in SymPy because `prime` does not accept symbolic inputs. 

Nominally to fix this for a given function, one just needs to take the function definition turn it into a `Function` subclass, then put the logic in the `eval` classmethod that returns None if the argument isn't an exact integer. But it would also be good to add things like pretty printing, and other symbolic methods or simplifications if possible.

**Comment by oscargus** [comment]
To clarify the printing part:

There should also be methods `_print_prime(self, expr)` added to (at least) LaTeX, pretty and MathML presentation printers that returns p with m subscripted for for `prime(m)`, so `p_{m}` for the LaTeX printer and `<msub><mo>p</mo><mi>m</mi></msub>` for MathML presentation. (An option is to add methods `_latex(self, printer=None)`, `_pretty(self, printer=None)` and `_mathml_presentation(self, printer=None)` to the class, or a combination of those.)

**Comment by czgdp1807** [comment]
IMO, this is an important issue.

**Comment by oscarbenjamin** [comment]
I think that there should be a separation between those functions in ntheory which are used for basic algorithms with concrete integers and those which are symbolic functions for use in a higher-level context. A symbolic `prime(n)` function would necessarily be something that is implemented on top of the core. Other functions in ntheory are used internally by the core e.g.:
https://github.com/sympy/sympy/blob/e33349b443b3c5b90e23af79f6995a69b592f0cd/sympy/core/numbers.py#L1979
https://github.com/sympy/sympy/blob/e33349b443b3c5b90e23af79f6995a69b592f0cd/sympy/core/numbers.py#L2319
https://github.com/sympy/sympy/blob/e33349b443b3c5b90e23af79f6995a69b592f0cd/sympy/core/numbers.py#L2407
https://github.com/sympy/sympy/blob/e33349b443b3c5b90e23af79f6995a69b592f0cd/sympy/core/mul.py#L918
https://github.com/sympy/sympy/blob/e33349b443b3c5b90e23af79f6995a69b592f0cd/sympy/core/mul.py#L1506
I would rather see an organisational separation such that the functions in ntheory are basic non-symbolic routines operating at a lower level that can be used e.g. by the core without risk of circular imports/definitions. The symbolic ntheory routines could be defined in the functions package and could also make use of the lower-level ntheory routines.

**Comment by asmeurer** [comment]
That sounds good, but I don't think it's necessary to expose the purely numeric functions to the user. The existing `prime` name (and so on) can be made symbolic, with automatic evaluation for integer arguments, which calls out to an internal `_prime` function that is only used for the symbolic `prime` and places in library code that know they will only be passing integers. 

Also, a lot of the functions should be analyzed to see if the make sense as symbolic versions. Something like `factorint` or `factorrat` don't really make sense symbolically, in my opinion. They return a dictionary, and even as returning an unevaluated factored Mul, they don't really make much sense for symbolic inputs. Similarly, something like `crt()` (Chinese Remainder Theorem) can be represented symbolically with sets. `isprime` is already a boolean via `Q.prime`, so I don't think it needs to be duplicated. `perfect_power` is also a boolean, so it should be part of the assumptions. It should already be possible to write a predicate with the existing assumptions that represents `perfect_power`, but maybe a shortcut assumption for it makes sense? `multinomial_coefficients_iterator` is an iterator and shouldn't be symbolic. 

In short, only the functions that return numbers should be symbolic, or the functions that return True or False, which should be predicates that are part of the assumptions system.

**Comment by oscarbenjamin** [comment]
> I don't think it's necessary to expose the purely numeric functions to the user

Maybe not but my point is really about the organisation of code internally in sympy. The fact that some parts of ntheory are used in core makes me think that those should be separated (in different modules) from any code that is built in top of the core. SymPy's problems with cyclic imports etc arise because the organisation of the dependencies across the codebase is not acyclic.

To me it makes sense to have a module called ntheory that contains non-symbolic routines whether they are for users or not. Of course we may want to have symbolic routines for users corresponding to the same mathematical functions but the code for those doesn't have to be in the ntheory package. An obvious place for such functions would be a `sympy.functions.ntheory` subpackage.
